### PR TITLE
Improve performance of select-top-k  WARP_SORT implementation

### PR DESCRIPTION
**Pull request by achirkin**
A few simplifications and tricks to improve the performance of the kernel:

- Promote some constants to static constexpr
- Allow `capacity < WarpSize`
- Reduce the frequency of `sort` operations for `filtered` version
- Remove `warp_sort::load` to simplify the api and implementation

### CODE REVIEW

**Comment by achirkin** [comment]
NB from previous discussions: I considered merging `done()` and `store(..)` together, but decided against that for two reasons:

  1. Atm, `store(...)` is rather fast and can be used to store the content multiple times, after `done()` has been called once. Looks like a useful behavior.
  2. I couldn't come up with a clean solution without the use of virtual functions, but I'm afraid introducing one would incur extra overheads.

**Comment by achirkin** [comment]
NB: more tests are coming in https://github.com/rapidsai/raft/pull/618

**Review by tfeher — CHANGES_REQUESTED** [review]
Hi Artem, thanks for the PR! It looks good, just a found a few small things. 

Question, is the following statement true?  With these modifications, capacity < warpsize is allowed. Since the capacity is chosen automatically according to the k value, the existing test cover these changes.

Please add to the PR description that the warp_sort::load functions are removed to simplify the API.

File: cpp/include/raft/spatial/knn/detail/topk/warpsort_topk.cuh [hunk 1]
```diff
@@ -151,60 +133,91 @@ class warp_sort {
...
+   *
+   * @param[in] in
+   *    a device pointer to a contiguous array, unique per-subwarp
+   *    (length: k <= kWarpWidth * kMaxArrLen).
+   * @param[in] in_idx
+   *    a device pointer to a contiguous array, unique per-subwarp
+   *    (length: k <= kWarpWidth * kMaxArrLen).
+   * @param[in] do_merge
+   *    servers as a conditional; when `false` the function does nothing;
+   *    it's needed to make sure threads within warp don't diverge when kWarpWidth < WarpSize.
```
[the remark this answers is not in the document]
Comment (achirkin) [reply cpp/include/raft/spatial/knn/detail/topk/warpsort_topk.cuh:176]: I'm referring there to `topk::bitonic(..).merge()`, which should run by all threads in a warp. In the code there, always only one of the following two is true: either `kWarpWidth < WarpSize` or the value of `do_merge` is the same for all threads in a warp.
Thanks for noting this, I'll update the documentation!

**Review by tfeher — APPROVED** [review]
Thanks Artem for addressing the issues! The PR looks good to me.

**Comment by cjnolet** [comment]
@gpucibot merge

**Pull request merged**